custom_model_runner/datarobot_drum/drum/drum.py

```diff
@@ -759,26 +775,65 @@ def _run_inside_docker(self, options, run_mode, raw_arguments):
 
     def _maybe_build_image(self, docker_image_or_directory):
         ret_docker_image = None
-        client = docker.client.from_env()
 
         if os.path.isdir(docker_image_or_directory):
             docker_image_or_directory = os.path.abspath(docker_image_or_directory)
-            self.logger.info(
-                "Building a docker image from directory {}...".format(docker_image_or_directory)
+            docker_build_msg = "Building a docker image from directory: {}...".format(
+                docker_image_or_directory
             )
+            self.logger.info(docker_build_msg)
             self.logger.info("This may take some time")
             try:
-                image, _ = client.images.build(path=docker_image_or_directory)
-                ret_docker_image = image.id
-            except docker.errors.BuildError as e:
-                self.logger.error("Hey something went wrong with image build!")
-                for line in e.build_log:
+                # Set image tag to the dirname of the docker context.
+                # This may be confusing if try to build different images from different contexts,
+                # with the same folder name:
+                # /path1/my_env
+                # /path2/my_env
+                #
+                # If image with the tag `my_env` exists, it will be untagged.
+                tag = os.path.basename(docker_image_or_directory)
+                client_docker_low_level = docker.APIClient()
+                spinner = Spinner(docker_build_msg + "  ")
+                json_lines = []
+                # Build docker, rotate spinner according to build progress
+                # and save status messages from docker build.
+                for line in client_docker_low_level.build(
+                    path=docker_image_or_directory, rm=True, tag=tag
```

Comment:
if you really wanted progress you could do something like find the total dockerfile lines and print progress per line that was built or someth

Reply:
Good idea....

Reply:
I tried this.
But drawback is that for a step `RUN pip install datarobot-drum` there is a lot of output: several lines for each pulled dependency.
For progressbar, to increment it properly, I need to estimate number of output lines, while for spinner - i just spin it on every output.

E.g:
3 steps in Dockerfile, each step outputs 100 lines: in case of progress bar I can only set it''s length to 3 and promote on every output, that contains `Step`.